In [3]:
import pandas as pd
import plotly.express as px
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.naive_bayes import MultinomialNB

# Import des données


In [4]:
df = pd.read_csv(
    "../SMSSpamCollection.txt", delimiter="\t", header=None, names=["spam", "text"]
)
df

spam                                               text
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

On cherche à modéliser la probabilité qu'un message soit un spam à partir de son contenu. Il nous faut donc un modèle de classification binaire. On peux donc utiliser un modèle linaire SVC, ou un naive bayes.

# Prétraitement des données

Je vais d'abord binariser la variable cible, sous forme d'un entier (0 ou 1).

In [5]:
df["spam"] = df["spam"].map({'spam': 1, 'ham': 0}).astype(int)


In [6]:
df

spam                                               text
0        0  Go until jurong point, crazy.. Available only ...
1        0                      Ok lar... Joking wif u oni...
2        1  Free entry in 2 a wkly comp to win FA Cup fina...
3        0  U dun say so early hor... U c already then say...
4        0  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567     1  This is the 2nd time we have tried 2 contact u...
5568     0               Will ü b going to esplanade fr home?
5569     0  Pity, * was in mood for that. So...any other s...
5570     0  The guy did some bitching but I acted like i'd...
5571     0                         Rofl. Its true to its name

[5572 rows x 2 columns]

# Approche par constructions de feature descriptives des messages

## Création des features

On vérifie la présence d'un numéro de téléphone.

In [7]:
df['has_phone_number'] = df['text'].apply(lambda x: 1 if re.search(r'\b\d{10,}\b', x) else 0)
df

spam                                               text  \
0        0  Go until jurong point, crazy.. Available only ...   
1        0                      Ok lar... Joking wif u oni...   
2        1  Free entry in 2 a wkly comp to win FA Cup fina...   
3        0  U dun say so early hor... U c already then say...   
4        0  Nah I don't think he goes to usf, he lives aro...   
...    ...                                                ...   
5567     1  This is the 2nd time we have tried 2 contact u...   
5568     0               Will ü b going to esplanade fr home?   
5569     0  Pity, * was in mood for that. So...any other s...   
5570     0  The guy did some bitching but I acted like i'd...   
5571     0                         Rofl. Its true to its name   

      has_phone_number  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  
...                ...  
5567                 1  
5568                 0  
5569                 0  
5570                 0  
5571                 0  

[5572 rows x 3 columns]

On vérifie la présence de monnaies.

In [8]:
df['has_currency_symbol'] = df['text'].apply(lambda x: 1 if re.search(r'[\$\€\£]', x) else 0)
df

spam                                               text  \
0        0  Go until jurong point, crazy.. Available only ...   
1        0                      Ok lar... Joking wif u oni...   
2        1  Free entry in 2 a wkly comp to win FA Cup fina...   
3        0  U dun say so early hor... U c already then say...   
4        0  Nah I don't think he goes to usf, he lives aro...   
...    ...                                                ...   
5567     1  This is the 2nd time we have tried 2 contact u...   
5568     0               Will ü b going to esplanade fr home?   
5569     0  Pity, * was in mood for that. So...any other s...   
5570     0  The guy did some bitching but I acted like i'd...   
5571     0                         Rofl. Its true to its name   

      has_phone_number  has_currency_symbol  
0                    0                    0  
1                    0                    0  
2                    0                    0  
3                    0                    0  
4                    0                    0  
...                ...                  ...  
5567                 1                    1  
5568                 0                    0  
5569                 0                    0  
5570                 0                    0  
5571                 0                    0  

[5572 rows x 4 columns]

On calcule la longueur des messages.

In [9]:
df['message_length'] = df['text'].apply(len)
df

spam                                               text  \
0        0  Go until jurong point, crazy.. Available only ...   
1        0                      Ok lar... Joking wif u oni...   
2        1  Free entry in 2 a wkly comp to win FA Cup fina...   
3        0  U dun say so early hor... U c already then say...   
4        0  Nah I don't think he goes to usf, he lives aro...   
...    ...                                                ...   
5567     1  This is the 2nd time we have tried 2 contact u...   
5568     0               Will ü b going to esplanade fr home?   
5569     0  Pity, * was in mood for that. So...any other s...   
5570     0  The guy did some bitching but I acted like i'd...   
5571     0                         Rofl. Its true to its name   

      has_phone_number  has_currency_symbol  message_length  
0                    0                    0             111  
1                    0                    0              29  
2                    0                    0             155  
3                    0                    0              49  
4                    0                    0              61  
...                ...                  ...             ...  
5567                 1                    1             160  
5568                 0                    0              36  
5569                 0                    0              57  
5570                 0                    0             125  
5571                 0                    0              26  

[5572 rows x 5 columns]

## Entrainement et évaluation du modèle

In [10]:

X = df[['message_length', 'has_phone_number', 'has_currency_symbol']]
y = df['spam']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LinearSVC()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.94      1.00      0.97       966
           1       0.97      0.61      0.75       149

    accuracy                           0.95      1115
   macro avg       0.96      0.80      0.86      1115
weighted avg       0.95      0.95      0.94      1115



# Approche par Vectorisation de text (NLP)

Il faut donc convertir le texte en une représentation numérique. J'utilise pour cela TfidfVectorizer.  C'est une technique de pondération des mots utilisée dans la récupération d'informations et le traitement du langage naturel (NLP). Elle évalue l'importance d'un mot dans un document par rapport à une collection de documents (corpus).
- TF (Term Frequency) : Nombre de fois qu'un mot apparaît dans un document, divisé par le nombre total de mots dans ce document.
- IDF (Inverse Document Frequency) : Mesure l'importance d'un mot en tenant compte de sa fréquence dans tout le corpus. Moins un mot est fréquent dans les autres documents, plus sa valeur IDF est élevée.

In [11]:
vectorizer = TfidfVectorizer(max_df=0.9, min_df=0.01)
X = vectorizer.fit_transform(df['text'])
y = df['spam']


## Utilisation du modèle naive bayes

On sépare les données d'entrainement et de tests.

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


On entraine le modèle.

In [13]:
model = MultinomialNB()
model.fit(X_train, y_train)


MultinomialNB()

On évalue le modèle sur les données de test.

In [14]:
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.9730941704035875
              precision    recall  f1-score   support

           0       0.97      1.00      0.98       966
           1       0.98      0.82      0.89       149

    accuracy                           0.97      1115
   macro avg       0.97      0.91      0.94      1115
weighted avg       0.97      0.97      0.97      1115



Avec la méthode de cross validation : 

In [15]:
print(cross_val_score(model, X, y, cv=3))

[0.96609257 0.96876683 0.96715132]


Le modèle est assez performant mais 25% des vrais spams n'ont pas été détectés (faux négatifs), donc non suffisant.

## Utilisation du modèle linéraire SVC

On sépare les données d'entrainement et de tests.

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


On entraine le modèle.

In [17]:
model = LinearSVC()
model.fit(X_train, y_train)


LinearSVC()

On évalue le modèle sur les données de test.

In [18]:
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


Accuracy: 0.9829596412556054
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       966
           1       0.95      0.92      0.94       149

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



Avec la méthode de cross validation :

In [19]:
print(cross_val_score(model, X, y, cv=3))

[0.97739505 0.9822294  0.97738288]


On obtient une précision plutot élevée, avec une accuracy de 98%, même si certains spams sont toujours manqués.